### ISSUE TITLE
Question: creating new styles (grooves)

### ISSUE THREAD

**Issue by lastant** [felixroos/jazzband#5]
How would you recommend to add other styles? Did you create existing styles by converting them from another source (e.g. Impro-Visor)?
Also, have you considered using midi style-files from keyboards (e.g. from http://www.jososoft.dk/yamaha/stylesites.htm)? This would allow a great diversity of styles.

**Comment by felixroos** [comment]
Hello!
Currently, styles are created using a custom object format. For example, here is a relatively simple "Funk" Style:

```ts
import { randomElement } from "../util/util";
export const funk = {
  name: 'Funk',
  tempo: 90,
  chords: () => randomElement([
    [[2, 0, 0, 1], 0, [0, .6], [0, 3.5, 0, 0]],
    [[1, 0, 0, 5], 0, [0, .6], [1, .5, 0, 0]],
    [[1, 0, 0, 5], 0, [0, .6], [.5, 0, 2, 0]],
    [[1, 0, 0, 5], 0, [0, .6], [0, 2, 0, 1]],
    [[3, 0, 0, 0], 0, 2, 0]
  ]),
  bass: () => randomElement(
    [
      [[2, 0, 0, 1], [0, 1], 1, [0, 1]],
      [[2, 0, 0, 0], [0, 1, 0, 1], [0, 0, 1, 0], [0, 0, 0, 1]],
    ]
  ),
  kick: (t) => [[[1, .7], 0, [0, 1], 0]],
  snare: (t) => [0, 1, [0, .2, 0, 0], [1, 0, 0, .6]],
  hihat: (t) => [[.5, .6], [.4, .7], [.3, .6], [.5, .7]]
};
```

You can find [all current grooves here](https://github.com/felixroos/jazzband/tree/master/src/grooves). I created them all by hand, with a little inspiration from ireal pro.

The groove object can be passed to the comp methods settings:

```js
band.comp(sheet, {
/* .. */
  groove: funk
});
```

In the band, each musician then plays another part of the groove, for example, the bassist will pick the "bass" value, as [implemented here](https://github.com/felixroos/jazzband/blob/master/src/musicians/Bassist.ts#L19).

This was only my first attempt in creating a playalong / groove library on the web. You might see it as a proof of concept. 

As I went deeper into this rabbit hole I decided to split the problems up into the [rhythmical](https://github.com/felixroos/rhythmical), the [harmonical](https://github.com/felixroos/harmonical) and the [leadsheet](https://github.com/felixroos/leetsheet) part (the latter being also influenced by Impro-Visor). When those implementations reach a certain state of stableness, I will reimplement jazzband with them. If you're interested, you can read about my current music coding explorations on my [blog](https://felixroos.github.io/blog/)

I did not yet look into keyboard style-files, but it looks worth checking out. I like the idea of having a huge set of data available, only needing to write an adapter. May I ask about how you came to this lib, and what interests you in this topic?

**Comment by lastant** [comment]
Yes, I saw the definitions of styles, but what is the structure behind those arrays under 'chords', 'bass' etc? I think your approach is similar to MMA, have you seen its code?
I'm developing a program that will need to have a playalong among other things, and I was looking at other projects to not reinvent the wheel. I want to use the ireal format, so your library fits very well. The only obstacle is the number of styles, but I can probably start with these four.

**Comment by lastant** [comment]
Besides, CASM styles have suitable intros and endings. These are currently missing..

**Comment by felixroos** [comment]
The arrays are what i call [nested rhythms](https://felixroos.github.io/blog/rhythmical-arrays/). The value of the primitives can generally be anything, in this case, the numbers are interpreted as durations of the notes, for example, the uppermost array:

```js
[[2, 0, 0, 1], 0, [0, .6], [0, 3.5, 0, 0]]
```

- it has 4 children elements which can be seen a measure with a subdivision of 4: each being 1/4 of the total duration
- those children elements of the measure can be subdivided again (if using an array)
- the children with 4 elements represent blocks of 16th notes: each (1/4)/4 = 1/16 of the total duration
- the child with 2 elements represent a block of 8th notes: each (1/4)/2 = 1/8 of the total duration

at some point, the nesting ends with a primitive, in this case with numbers. They can be multiplied with their duration fraction:

- the duration of the "2" is 1/(4/4) = 1/16 * 2 = 1/8
- the duration of the "1" is 1/(4/4) = 1/16 * 1 = 1/16
- the duration of the ".6" is 1/(4/2) = 1/8 * 0.6 = a shorter 8th note
- the duration of the "3.5" is 1/(4/4) = 1/16 * 3.5 = 3+1/2 16th notes (will leap over to next bar => not the best choice here) 
- the zeroes act as rests

I have written some more in depth blog posts [about the nested arrays](https://felixroos.github.io/blog/rhythmical-arrays/). But keep in mind that the jazzband lib is just using the nested idea, not the implementation described there. The core rhythm mechanic of this lib is [arrayPulse](https://github.com/felixroos/jazzband/blob/master/src/Pulse.ts#L26), which is called by each musician using [tickArray](https://github.com/felixroos/jazzband/blob/master/src/Pulse.ts#L71).

Intros and endings could potentially work as part of the groove if you make use of the [function params](https://github.com/felixroos/jazzband/blob/master/src/musicians/Pianist.ts#L51) [passed to each groove track](https://github.com/felixroos/jazzband/blob/master/src/musicians/Bassist.ts#L29), checking if the current measure is first or last to return a different pattern etc.. I have not tried this yet, but I would be interested if it works for you. 

As I'm currently reimplementing all of this and more, I don't see myself adding new features as part of this lib before the reimplementation is done (idk how long it takes). But if the MMA styles are parseable, I would directly try them out with the new implementation. I have not seen any the MMA code yet, and I am not really into this keyboard stuff, maybe you could give me a more precise link or description?

**Comment by lastant** [comment]
MMA is a library similar to yours, but written in Python and generates a MIDI file instead of actually playing the music. Their approach to the styles is very similar: each style has a logical description file.

Yamaha styles (CASM) are basically MIDI files with a bit of extra info (CASM section): which part is which arrangement section (intro version 1, intro version 2, A, B, C - with optional versions, ending), so that the program can take different sections from the MIDI, adjust them according to the current key and compile a full arrangement. I like this approach, because it really makes sense and it allows easy creation of even complex styles with many instruments.

**Comment by felixroos** [comment]
I suppose you mean [this](https://www.mellowood.ca/mma/online-docs/html/tut/node3.html)? Sounds good, I only think the format looks a bit verbose. Not sure how they handle voicings or if they support melodies. It looks a little bit laborious to use though. Before I will be able to implement a proper MIDI/CASM parser, some foundation work has yet to be done.
Currently, my goal is to implement a good and flexible format that can be used for many musical applications like backing tracks, algorithmic (live) composition or ear training. I think i found a good way to handle polyphony and macro time control with [rhythmical objects](https://felixroos.github.io/blog/rhythmical-objects/) which were mainly inspired by [tidal cycles mini notation](https://tidalcycles.org/index.php/Tutorial) (nested notation) and [this paper](http://recherche.ircam.fr/equipes/repmus/LispSecondLanguage/functional/functional.html) (polyphony). 

At the moment, my effort goes into developing reducers that run on the generated flat events to add functionality, without cluttering the core format. I recently implemented [voicings](https://felixroos.github.io/blog/rhythmical-chords/) that can be generated from chord symbols, using pluggable voicing dictionaries. I also experimented with dynamic voicing generation (similar to Impro-Visor), which I will add soon. The next steps then will be relative pitch notation using degrees like "1 b3 5" (relative to a tonic) or scale steps like "1 3 5" (relative to tonic + scale). After that, I will be able to plug that all together into grooves, forming the basis to create CASM like tracks or even be able to generate custom arrangements by certain rules. Other plans involve melody harmonization, a domain specific language for nested notation, live midi input, live midi send, microtime adjustems (swing), rhythm permutation ... I could go on forever :D

**Comment by lastant** [comment]
Btw, just found this project that makes use of CASM styles: https://github.com/jjazzboss/JJazzLab-X

**Comment by felixroos** [comment]
I stopped working on this lib, as I moved to developing [strudel](https://github.com/tidalcycles/strudel), which is similar (but much more powerful)
